internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
+type innitInvocationContext interface {
+	InvocationContext
+	// Review: the spec has no error on this? what if the addr is already taken?
+	CreateActor(cid cid.Cid, addr address.Address, constructorParams MethodParams)
+}
+
+// Storage2 defines the storage module exposed to actors.
+// Review: check the old `Storage` for some more comments on why this looks like it does
+type Storage2 interface {
+	Put(interface{}) (cid.Cid, error)
+	Get(cid.Cid, interface{}) error
+}
+
+// Epoch is the current chain epoch.
+// Dragons: this is just to match names on the VM, Height got renamed to Epoch on the spec
+// TODO: rename types.BlockHeight to block.Epoch (issue: ???)
+type Epoch types.BlockHeight
+
+// Randomness is a string of random bytes
+// Review: placeholder for matching newtypes with the spec.
+type Randomness []byte
+
+// InvocInput are the params to invoke a method in an Actor.
+// Review: why is this a struct instead of arguments?
+type InvocInput struct {
+    To      address.Address
+    Method  types.MethodID
+    Params  MethodParams
+    Value   types.AttoFIL
+}
+
+// ExitCode is the exit code of a method executing inside the VM.
+// Review: do we have reserved codes?
+// TODO: this has type UVarint in the spec
+type ExitCode types.Int64
+
+// InvocOutput is waht os returned by an actor executing a method.
+// Review: multiple returns vs a return struct. There are some minor implications with swapping vm implementations written in another language.
+type InvocOutput struct {
+	ExitCode     ExitCode
+    ReturnValue  []byte
+}
+
+// MethodParam is the parameter to an Actor method already encoded for execution in the VM.
+type MethodParam []byte
+// MethodParams is a list of `MethodParam` to be passed into an Actor method.
+type MethodParams []MethodParam
+
 // Runtime defines the ABI interface exposed to actors.
 type Runtime interface {
+	// Review: this is not in the spec anymore
 	Message() *types.UnsignedMessage
+
+	// Review: the spec splits this up and brings the get/put to be first class
 	Storage() Storage
+
+	// Review: changed from arguments to a struct
 	Send(to address.Address, method types.MethodID, value types.AttoFIL, params []interface{}) ([][]byte, uint8, error)
+
+	// Review: this is not in the spec anymore
 	AddressForNewActor() (address.Address, error)
+
+ 	// Review: renamed
 	BlockHeight() *types.BlockHeight
+
+	// Review: this is not in the spec anymore
 	MyBalance() types.AttoFIL
+
+	// Review: this changed a bit, not sure why it cant be determined off the result of `Caller()`
 	IsFromAccountActor() bool
+
+	// Review: this is not in the spec anymore
 	Charge(cost types.GasUnits) error
+
+	// Review: this got renamed and gained an offset arg
 	SampleChainRandomness(sampleHeight *types.BlockHeight) ([]byte, error)
 
+	// unchanged
 	CreateNewActor(addr address.Address, code cid.Cid, initalizationParams interface{}) error
 
+	// Review: this is not in the spec anymore
```

Comment:
The spec's direct import in the miner code (`new(proving.StorageProvingSubsystem_I).VerifySeal(&sector.SealVerifyInfo_I{`) seems worse for testing and like something we can do our own way.

Reply:
I don't think this is fundamental; `verifySeal` is intended to be a pure function, but beyond that, it shouldn't matter how the source modules are split up

Reply:
I think the importance of that is that if it's not a runtime hook, it then is expected to be either: VM code, or the VM needs an ABI for calling external functions defined. 

Right now "VM code" happens to be happy Go, and calling some other happy random Go code seems ok, and we are not thinking on when we are crossing VM boundaries that much. But it has consequences when swapping VM implementations (specially ones written in another language) or when getting an actual VM defined with its own language.

Reply:
I see, that makes sense. It would be nice if these pure functions (and, in fact, all functions involving nontrivial computation) were proxied through the runtime, since this would make gas accounting much easier. On the other hand, we don't want to expand the runtime API surface without limit. Would it make sense to have a single runtime API method (e.g., `Call`) that takes an ID of a pure function, along with arguments (either serialized or with some standard ABI)?